File: csharp/AppEncryption/AppEncryption.IntegrationTests/Utils/PersistenceFactory.cs
using System.Collections.Concurrent;
using GoDaddy.Asherah.AppEncryption.Persistence;
using LanguageExt;

namespace GoDaddy.Asherah.AppEncryption.IntegrationTests.Utils
{
    public static class PersistenceFactory<T>
    {
        public static Persistence<T> CreateInMemoryPersistence()
        {
            return new DictionaryPersistence();
        }

        private class DictionaryPersistence : Persistence<T>
        {
            private readonly ConcurrentDictionary<string, T> dictionaryPersistence = new ConcurrentDictionary<string, T>();

            public override Option<T> Load(string key)
            {
                return dictionaryPersistence.TryGetValue(key, out T result) ? result : Option<T>.None;
            }

            public override void Store(string key, T value)
            {
                dictionaryPersistence[key] = value;
            }
        }
    }
}

File: java/app-encryption/src/it/java/com/godaddy/asherah/utils/PersistenceFactory.java
package com.godaddy.asherah.utils;

import com.godaddy.asherah.appencryption.persistence.Persistence;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

public final class PersistenceFactory {

  private PersistenceFactory() {
  }

  public static <T> Persistence<T> createInMemoryPersistence() {
    return new Persistence<T>() {
      private final Map<String, T> mapPersistence = new ConcurrentHashMap<>();

      @Override
      public Optional<T> load(final String key) {
        return Optional.ofNullable(mapPersistence.get(key));
      }

      @Override
      public void store(final String key, final T value) {
        mapPersistence.put(key, value);
      }
    };
  }
}
